plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -140,23 +139,36 @@ public StatisticsFile writeStatisticsFile(
                 table.statisticsFiles().stream()
                         .filter(statisticsFile -> statisticsFile.snapshotId() == snapshotId)
                         .collect(toOptional())
-                        .ifPresent(previousStatisticsFile -> copyRetainedStatistics(fileIO, previousStatisticsFile, validFieldIds, ndvSketches.keySet(), writer));
+                        .ifPresent(previousStatisticsFile -> copyRetainedStatistics(fileIO, previousStatisticsFile, validFieldIds, merged.keys(), writer));
 
-                ndvSketches.entrySet().stream()
+                merged.getAllEntries()
                         .sorted(comparingByKey())
                         .forEachOrdered(entry -> {
                             Integer fieldId = entry.getKey();
-                            CompactSketch sketch = entry.getValue();
-                            @SuppressWarnings("NumericCastThatLosesPrecision")
-                            long ndvEstimate = (long) sketch.getEstimate();
-                            writer.add(new Blob(
-                                    APACHE_DATASKETCHES_THETA_V1,
-                                    ImmutableList.of(fieldId),
-                                    snapshotId,
-                                    snapshotSequenceNumber,
-                                    ByteBuffer.wrap(sketch.toByteArray()),
-                                    ZSTD,
-                                    ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, Long.toString(ndvEstimate))));
+                            Object value = entry.getValue();
+                            if (value instanceof CompactSketch sketch) {
+                                @SuppressWarnings("NumericCastThatLosesPrecision")
+                                long ndvEstimate = (long) sketch.getEstimate();
+                                writer.add(new Blob(
+                                        APACHE_DATASKETCHES_THETA_V1,
+                                        ImmutableList.of(fieldId),
+                                        snapshotId,
+                                        snapshotSequenceNumber,
+                                        ByteBuffer.wrap(sketch.toByteArray()),
+                                        ZSTD,
+                                        ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, Long.toString(ndvEstimate))));
+                            }
+                            else {
```

Comment:
per the comment elsewhere, the `else {` sholdn't be needed, the old values should be taken care of using `copyRetainedStatistics`